### ISSUE TITLE
omnisharp in C# extension not finding mono library

### ISSUE THREAD

**Issue by csgo-react** [flathub/com.visualstudio.code#241]
At the visual studio output I get this error message:

```
OmniSharp server started.
    Path: /home/username/.var/app/com.visualstudio.code/data/vscode/extensions/ms-dotnettools.csharp-1.23.13/.omnisharp/1.37.12/run

System.TypeInitializationException: The type initializer for 'System.ComponentModel.TypeDescriptor' threw an exception. ---> System.TypeInitializationException: The type initializer for 'Sys' threw an exception. ---> System.DllNotFoundException: /home/username/.var/app/com.visualstudio.code/data/vscode/extensions/ms-dotnettools.csharp-1.23.13/.omnisharp/1.37.12/lib/../lib/libmono-native.so assembly:<unknown assembly> type:<unknown type> member:(null)
  at (wrapper managed-to-native) Interop+Sys.LChflagsCanSetHiddenFlag()
  at Interop+Sys..cctor () [0x00000] in <9f0df102fe6e4cfea29d2e46f585d8a5>:0 
   --- End of inner exception stack trace ---
  at Interop.GetRandomBytes (System.Byte* buffer, System.Int32 length) [0x00000] in <9f0df102fe6e4cfea29d2e46f585d8a5>:0 
  at System.Guid.NewGuid () [0x00000] in <9f0df102fe6e4cfea29d2e46f585d8a5>:0 
  at System.ComponentModel.TypeDescriptor..cctor () [0x00034] in <3d97dda773a54430b892e1d794e5d921>:0 
   --- End of inner exception stack trace ---
  at OmniSharp.Internal.CommandOptionExtensions.GetValueOrDefault[T] (McMaster.Extensions.CommandLineUtils.CommandOption opt, T defaultValue) [0x00008] in <69b931dd99474d7d81480d2726eb8b81>:0 
  at OmniSharp.Stdio.StdioCommandLineApplication.get_Encoding () [0x00000] in <b3195b0d6aee4d87a648926211f2a421>:0 
  at OmniSharp.Stdio.Driver.Program+<>c__DisplayClass0_1.<Main>b__1 () [0x00000] in <f1955cbf5725470fa824f268e783b3d1>:0 
  at OmniSharp.CommandLineApplication+<>c__DisplayClass11_0.<OnExecute>b__0 () [0x0000b] in <69b931dd99474d7d81480d2726eb8b81>:0 
  at McMaster.Extensions.CommandLineUtils.CommandLineApplication.Execute (System.String[] args) [0x00039] in <344a1e8cd6594b17b1e01f8df6ae8827>:0 
  at OmniSharp.CommandLineApplication.Execute (System.String[] args) [0x00081] in <69b931dd99474d7d81480d2726eb8b81>:0 
  at OmniSharp.Stdio.Driver.Program+<>c__DisplayClass0_0.<Main>b__0 () [0x00028] in <f1955cbf5725470fa824f268e783b3d1>:0 
  at OmniSharp.HostHelpers.Start (System.Func`1[TResult] action) [0x0001c] in <69b931dd99474d7d81480d2726eb8b81>:0 
[ERROR] Error: OmniSharp server load timed out. Use the 'omnisharp.projectLoadTimeout' setting to override the default delay (one minute).
```

I reproduced it again with:
* Installing Linux Mint 20.2 in Virtual Box and update all to current
* apt-install mono-complete
* ./dotnet-install -c 5.0
* installing Flatpak VSC at in Linux Mint Software manger (its v1.57.1)
* to work I added a $HOME/.dotnet to the PATH variable. Did this with Flatseal. (DOTNET_CLI_TELEMETRY_OPTOUT=1 as well, no idea if working at Linux)
* (In VSC I disabled all telemetry stuff)
* In VSC I installed C# extension
* In VSC I opened a folder and created a new project in console: dotnet new console
* After switching to 'Program.cs' (OmniSharp installed something at the very first) and OmniSharp throwing that error from above.

Doing the same with the official not-flatpak version did work. However I would prefer using it inside a flatpak

**Comment by mkfrey** [comment]
Changes to your host system do not have any effect inside flatpak containers, so running `apt-install mono-complete` and `./dotnet-install -c 5.0` is pointless. If you want to use dotnet, you can install the `org.freedesktop.Sdk.Extension.dotnet` and override the environment variable `FLATPAK_ENABLE_SDK_EXT` to include `dotnet` (`flatpak override --env='FLATPAK_ENABLE_SDK_EXT=dotnet'` to enable the installed dotnet SDK or `flatpak override --env='FLATPAK_ENABLE_SDK_EXT=*'` to simply enable all installed SDK extensions).

Now to your main problem: It is probably caused by libmono-native.so being dynamically linked to libkrb5.so.3 and other library files of krb5, which are not present in the flatpak container. This causes a failure when OmniSharp tries to load it.

How to resolve:
1) Download the source code of the latest release of krb5 [here](https://web.mit.edu/kerberos/dist/)
2) Extract the files
3) Enter the src directory inside the newly extracted folder and execute `./configure`. **This has to be done from inside the flatpak container using the VSCode commandline.**
3) `make`  (**VSCode commandline.**)
4) Override the LD_LIBRARY_PATH environment variable to include the `lib` folder, e.g. `flatpak override --user LD_LIBRARY_PATH=/home/user/Documents/Code/ZY_bin/krb5-1.19.1/src/lib`

**Comment by csgo-react** [comment]
Thanks for the quick response. This error is gone! Thanks for that!

And sorry I missed some important details. 
As in #242 my libmono-native.so is present as well.
Also my dotnet did work (after adding `$HOME/.dotnet `to the `PATH` variable). I could create new projects and build/run them as well.

I tested the alternative and installed org.freedesktop.Sdk.Extension.dotnet5 (need 5.0) with `env='FLATPAK_ENABLE_SDK_EXT=dotnet5'` . This also worked fine.

As you guessed correctly for `ldd libmono-native.so` the library missing some reference for `libgssapi_krb5.so.2` (not .3)

Downloading the linked source and running `configure` and `make` did the job.
Thanks for that!

---------------------------------------------------------------------------------------
But could there be an alternative solution? krb5 is also present at my main system. Is there any chance to use this (as I did for  `.dotnet` above) or another Flatpak extension?

Downloading cryptographic code from a website is a little itchy. So far I didn't manage to successfully checking the signature. 

---------------------------------------------------------------------------------------

Still not working is code completion. But that's a new issue.

**Comment by mkfrey** [comment]
See my comment in issue #242 on how to work around the OmniSharp completion.

Whether you can use your host krb5, I don't know. I think it would be a better idea to try getting the maintainers of org.freedesktop.Sdk to include it.

**Comment by mkfrey** [comment]
I created a pull request #243 to include Kerberos.

**Comment by csgo-react** [comment]
I though its something different in #242 because it did showed some documentation about stuff I build at least once. Sorry for that trouble, thanks again for helping. I think it's working now.

I tested different ways:
`flatpak update --user --commit=9b417d8010efc85484e9780a76277b213b8b0ef69fb6e5ee70591c411ec305ce com.visualstudio.code`  and `--no-sandbox` did work.
To avoid 'no sandbox' I downgraded it to 1.56.2 with
`flatpak update --user --commit=dab8194db81dd9a593dacb1d484457a6ba2585787c7ee90f2a6ce417239dc9ef com.visualstudio.code` ( #223 )
This worked without the  `--no-sandbox` parameter but the C# extension need to be installed again. 

Also tested the Kerberos test build. This worked with the krb5 problem (but also had not code completion).
